server/Database.hs

```diff
@@ -0,0 +1,187 @@
+    ()
+  Sqlite.execute
+    conn
+    " create table if not exists attested_credential_data                      \
+    \ ( id               blob    primary key                                   \
+    \ , user_id          blob    not null                                      \
+    \ , public_key_x     blob    not null                                      \
+    \ , public_key_y     blob    not null                                      \
+    \ , created          text    not null                                      \
+    \                    default (strftime('%Y-%m-%dT%H:%M:%fZ', 'now'))       \
+    \ , foreign key (user_id) references users (id)                            \
+    \ );                                                                       "
+    ()
+  Sqlite.execute
+    conn
+    " create index if not exists                                               \
+    \ ix_attested_credential_data_user_id                                      \
+    \ on attested_credential_data(user_id);                                    "
+    ()
+
+-- | Run an action using `Sqlite.withTransaction`.
+--
+-- If exceptions occur within the transaciton, the transaction is aborted
+-- with `ROLLBACK TRANSACTION`. Otherwise, the transaction is committed using
+-- `COMMIT TRANSACTION`.
+--
+-- This ensures that we always close our transactions and don't leave them
+-- open when exceptions occur.
+withTransaction :: Sqlite.Connection -> (Transaction -> IO a) -> IO a
+withTransaction conn action = Sqlite.withTransaction conn (action (Transaction conn))
+
+begin :: Sqlite.Connection -> IO Transaction
+begin conn = do
+  Sqlite.execute conn "begin;" ()
+  pure $ Transaction conn
+
+commit :: Transaction -> IO ()
+commit (Transaction conn) = Sqlite.execute conn "commit;" ()
+
+rollback :: Transaction -> IO ()
+rollback (Transaction conn) = Sqlite.execute conn "rollback;" ()
+
+addUser ::
+  Transaction ->
+  Fido2.PublicKeyCredentialUserEntity ->
+  IO ()
+addUser (Transaction conn) user =
+  let Fido2.PublicKeyCredentialUserEntity
+        { id = (UserId (URLEncodedBase64 userId)),
+          name = username,
+          displayName = displayName
+        } = user
+   in Sqlite.execute
+        conn
+        "insert into users (id, username, display_name) values (?, ?, ?);"
+        (userId, username, displayName)
+
+addAttestedCredentialData ::
+  Transaction ->
+  Fido2.UserId ->
+  Fido2.CredentialId ->
+  Fido2.PublicKey ->
+  IO ()
+addAttestedCredentialData
+  (Transaction conn)
+  (UserId (URLEncodedBase64 userId))
+  (CredentialId (URLEncodedBase64 credentialId))
+  publicKey =
+    Sqlite.execute
+      conn
+      " insert into attested_credential_data                        \
+      \ (id, user_id, public_key_x, public_key_y)                   \
+      \ values                                                      \
+      \ (?, ?, ?, ?);                                               "
+      ( credentialId,
+        userId,
```

Comment:
How about we define our own `PublicKey` datatype domain object that is an `x/y` pair and only convert to `ECDSA.Point` on demand when checking the signature. This way we don't need these partial `publicKeyX` functions.

Also it decouples storage from the crypto-implementation; which might help with #22

Reply:
The injection `OurPoint -> ECDSA.Point` will be total, as you never have the `Infinity` case.  

And then you can call `verify` or `sign` or whatever using the constructed point

Reply:
I've implemented this now.